### PR TITLE
DDNet network character

### PR DESCRIPTION
**Pull request by H-M-H**
could fix: #198 #258 
would be useful for: #262 #277

### CODE REVIEW

File: src/game/client/gameclient.cpp [hunk 1]
```diff
@@ -1041,18 +1041,39 @@ void CGameClient::OnNewSnapshot()
...
 			{
 				const void *pOld = Client()->SnapFindItem(IClient::SNAP_PREV, NETOBJTYPE_CHARACTER, Item.m_ID);
-				m_Snap.m_aCharacters[Item.m_ID].m_Cur = *((const CNetObj_Character *)pData);
+				m_Snap.m_aCharacters[Item.m_ID].m_Cur.CNetObj_Character::operator=(*((const CNetObj_Character *)pData));
 				if(pOld)
 				{
 					m_Snap.m_aCharacters[Item.m_ID].m_Active = true;
-					m_Snap.m_aCharacters[Item.m_ID].m_Prev = *((const CNetObj_Character *)pOld);
+					m_Snap.m_aCharacters[Item.m_ID].m_isDDNet = false;
+					m_Snap.m_aCharacters[Item.m_ID].m_Prev.CNetObj_Character::operator=(*((const CNetObj_Character *)pOld));
```
Comment (CytraL) [inline src/game/client/gameclient.cpp]: Why this change?

File: src/game/client/gameclient.cpp [hunk 2]
```diff
@@ -1041,18 +1041,39 @@ void CGameClient::OnNewSnapshot()
...
 			}
+			else if(Item.m_Type == NETOBJTYPE_CHARACTER_DDNET)
+			{
+				const void *pOld = Client()->SnapFindItem(IClient::SNAP_PREV, NETOBJTYPE_CHARACTER_DDNET, Item.m_ID);
+				m_Snap.m_aCharacters[Item.m_ID].m_Cur = *((const CNetObj_Character_DDNet *)pData);
+				if(pOld)
+				{
+					m_Snap.m_aCharacters[Item.m_ID].m_Active = true;
+					m_Snap.m_aCharacters[Item.m_ID].m_isDDNet = true;
+					m_Snap.m_aCharacters[Item.m_ID].m_Prev = *((const CNetObj_Character_DDNet *)pOld);
```
Comment (CytraL) [inline src/game/client/gameclient.cpp]: And why not here?

Comment (H-M-H) [inline src/game/client/gameclient.cpp]: Here we received a complete CNetObj_Character_DDNet and can directly assign it.

**Comment by H-M-H** [comment]
What this does:

if a DDNet-Client with proper netversion connects, the server sends its netversion back. Compatibility is made sure like that.

If client and server match the server will send more detailed character infos, to be more precise it will send a ddnet specific character-netobject that inherits from the standard netobject. This can be done in the same way for other netobjects to add additional ddnet-specific infos.

This is quite old, is there some better solution you can think of now or should I have a look at this again ?
